ragna/core/_components.py

```diff
@@ -42,30 +50,47 @@ def __repr__(self) -> str:
     def _protocol_models(
         cls,
     ) -> dict[tuple[Type[Component], str], Type[pydantic.BaseModel]]:
+        # This method dynamically builds a pydantic.BaseModel for the extra parameters
+        # of each method that is listed in the __ragna_protocol_methods__ class
+        # variable. These models are used by ragna.core.Chat._unpack_chat_params to
+        # validate and distribute the **params passed by the user.
+
+        # Walk up the MRO until we find the __ragna_protocol_methods__ variable. This is
+        # the indicator that we found the protocol class. We use this as a reference for
+        # which params of a protocol method are part of the protocol (think positional
+        # parameters) and which are requested by the concrete class (think keyword
+        # parameters).
         protocol_cls, protocol_methods = next(
             (cls_, cls_.__ragna_protocol_methods__)  # type: ignore[attr-defined]
             for cls_ in cls.__mro__
             if "__ragna_protocol_methods__" in cls_.__dict__
         )
         models = {}
         for method_name in protocol_methods:
+            num_protocol_params = len(
+                inspect.signature(getattr(protocol_cls, method_name)).parameters
+            )
             method = getattr(cls, method_name)
-            concrete_params = inspect.signature(method).parameters
-            protocol_params = inspect.signature(
-                getattr(protocol_cls, method_name)
-            ).parameters
-            extra_param_names = concrete_params.keys() - protocol_params.keys()
-
-            models[(cls, method_name)] = pydantic.create_model(  # type: ignore[call-overload]
+            params = iter(inspect.signature(method).parameters.values())
+            annotations = get_type_hints(method)
+            # Skip over the protocol parameters in order for the model below to only
```

Comment:
This is the main thing that was biting me when trying to add new required parameters to methods on `Assistant` subclasses. This seems to resolve everything for me, although full disclosure I haven't been able to think through all the nuances of how this might affect custom user components.